src/main/java/io/vertx/ext/httpservicefactory/HttpServiceFactory.java

```diff
@@ -0,0 +1,388 @@
+    });
+    req.handler(resp -> {
+      int status = resp.statusCode();
+      if (status == 200) {
+        resp.pause();
+        vertx.fileSystem().open(file.getPath(), new OpenOptions().setCreate(true), ar2 -> {
+          if (ar2.succeeded()) {
+            resp.resume();
+            AsyncFile result = ar2.result();
+            String contentType = resp.getHeader("Content-Type");
+            int index = contentType.indexOf(";");
+            String mediaType = index > -1 ? contentType.substring(0, index) : contentType;
+            Function<Buffer, Buffer> unmarshaller = unmarshallerFactory.apply(mediaType);
+            Pump pump = Pump.pump(resp, new WriteStream<Buffer>() {
+              @Override
+              public WriteStream<Buffer> exceptionHandler(Handler<Throwable> handler) {
+                result.exceptionHandler(handler);
+                return this;
+              }
+              @Override
+              public WriteStream<Buffer> write(Buffer data) {
+                data = unmarshaller.apply(data);
+                if (data != null) {
+                  result.write(data);
+                }
+                return this;
+              }
+              @Override
+              public WriteStream<Buffer> setWriteQueueMaxSize(int maxSize) {
+                result.setWriteQueueMaxSize(maxSize);
+                return this;
+              }
+              @Override
+              public boolean writeQueueFull() {
+                return result.writeQueueFull();
+              }
+              @Override
+              public WriteStream<Buffer> drainHandler(Handler<Void> handler) {
+                result.drainHandler(handler);
+                return this;
+              }
+            });
+            pump.start();
+            resp.exceptionHandler(err -> {
+              handler.handle(Future.failedFuture(ar2.cause()));
+            });
+            resp.endHandler(v1 -> {
+              Buffer last = unmarshaller.apply(null);
+              if (last != null) {
+                result.write(last);
+              }
+              result.close(v2 -> {
+                if (v2.succeeded()) {
+                  handler.handle(Future.succeededFuture(file));
+                } else {
+                  handler.handle(Future.failedFuture(v2.cause()));
+                }
+              });
+            });
+          } else {
+            handler.handle(Future.failedFuture(ar2.cause()));
+          }
+        });
+      } else if (status == 401) {
+        if (prefix().equals("https") && resp.getHeader("WWW-Authenticate") != null && username != null && password != null) {
+          doRequest(client, file, url, username, password, true, unmarshallerFactory, handler);
+          return;
+        }
+        handler.handle(Future.failedFuture(new Exception("Unauthorized")));
+      } else {
+        handler.handle(Future.failedFuture(new Exception("Cannot get file status:" + status)));
+      }
```

Comment:
We should handle redirects too.

I'd recommend looking at the code from Vert.x 2 - you should be able to copy and paste most of it.